sourcegraph/sourcegraph
# Redirect to a specified returnTo path after the post-sign-up survey

The post-sign up survey hard-codes a redirect to `/get-cody`. To simplify getting a Cody token from dotcom in fresh Cody VScode installs, we plan to make our login flow:

1. Login (GitHub, GitLab, email) with a redirect to the post-sign up survey.
2. Post sign up survey, with a redirect to issue tokens.
3. Issue token, with a redirect to the VScode extension.

This change enables step 2, by changing the post-sign up survey redirect to a query string parameter.

## Test plan

Automated tests:

```
cd client/web
pnpm test -- -t '.*PostSignUpPage.*'
```

Manual test:

0. `sg start dotcom`
1. `sg db add-user --username test2 --password test2`
2. https://sourcegraph.test:3443/sign-in?showMore=&returnTo=/post-sign-up?returnTo=/user/settings/tokens/new/callback?requestFrom=CODY
3. Sign in as test2/test2; you should see the post sign-up survey.
4. Complete the survey, you should see the prompt to "Authorize Cody AI by Sourcegraph - VS Code Extension"
5. Click Authorize, browser should try to open a vscode://sourcegraph.cody-ai link, may prompt you if you have VScode installed.
6. Case where the user has completed post sign-up already: https://sourcegraph.test:3443/sign-in?showMore=&returnTo=/post-sign-up?returnTo=/user/settings/tokens/new/callback?requestFrom=CODY
7. You should arrive directly at the "Authorize Cody AI by Sourcegraph - VS Code Extension" prompt.

The same redirects should work for GitHub and GitLab authorization URLs, but you can't test these with the `sourcegraph.test` domain.

Note you can look at a user's email validation/post sign-up bits in the admin console:

```
query {
  user(username: "test2") {
    username,
    hasVerifiedEmail,
    completedPostSignup
  }
}
```
